File: golang/tests/pkg/unit/filesystem_watch_ws_push_test.go
package agentbay_test

import (
	"fmt"
	"sync"
	"testing"
	"time"

	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/filesystem"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
)

// MockWsSession implements both the FileSystem.Session interface and the
// wsProvider optional interface so that WatchDirectory can attempt WS push.
type MockWsSession struct {
	mock.Mock
}

func (m *MockWsSession) GetAPIKey() string {
	args := m.Called()
	return args.String(0)
}

func (m *MockWsSession) GetClient() *mcp.Client {
	args := m.Called()
	if args.Get(0) == nil {
		return nil
	}
	return args.Get(0).(*mcp.Client)
}

func (m *MockWsSession) GetSessionId() string {
	args := m.Called()
	return args.String(0)
}

func (m *MockWsSession) CallMcpTool(toolName string, args interface{}) (*models.McpToolResult, error) {
	mockArgs := m.Called(toolName, args)
	return mockArgs.Get(0).(*models.McpToolResult), mockArgs.Error(1)
}

func (m *MockWsSession) GetWsUrl() string {
	args := m.Called()
	return args.String(0)
}

func (m *MockWsSession) GetToken() string {
	args := m.Called()
	return args.String(0)
}

func (m *MockWsSession) GetWsClient() (interface{}, error) {
	args := m.Called()
	return args.Get(0), args.Error(1)
}

func (m *MockWsSession) GetMcpServerForTool(toolName string) string {
	args := m.Called(toolName)
	return args.String(0)
}

// Logger interface stubs
func (m *MockWsSession) LogInfo(msg string)  {}
func (m *MockWsSession) LogDebug(msg string) {}
func (m *MockWsSession) LogWarn(msg string)  {}
func (m *MockWsSession) LogError(msg string) {}

// TestWsPush_FallsBackToPollingWhenNoWsUrl verifies that when WsUrl is empty,
// WatchDirectory falls back to polling mode.
func TestWsPush_FallsBackToPollingWhenNoWsUrl(t *testing.T) {
	sess := new(MockWsSession)
	fs := filesystem.NewFileSystem(sess)

	sess.On("GetWsUrl").Return("")
	sess.On("GetToken").Return("test-token")
	sess.On("GetMcpServerForTool", "get_file_change").Return("wuying_filesystem")

	pollCount := 0
	var mu sync.Mutex
	sess.On("CallMcpTool", "get_file_change", map[string]string{"path": "/tmp/poll"}).
		Return(&models.McpToolResult{Success: true, Data: "[]"}, nil).
		Run(func(args mock.Arguments) {
			mu.Lock()
			pollCount++
			mu.Unlock()
		})

	stopCh := make(chan struct{})
	wg, readyCh := fs.WatchDirectory("/tmp/poll", func([]*filesystem.FileChangeEvent) {}, 50*time.Millisecond, stopCh)
	<-readyCh

	time.Sleep(150 * time.Millisecond)
	close(stopCh)
	wg.Wait()

	mu.Lock()
	assert.GreaterOrEqual(t, pollCount, 2, "Should have polled at least twice (baseline + ticker)")
	mu.Unlock()
}

// TestWsPush_FallsBackToPollingWhenNoServerName verifies that when
// GetMcpServerForTool returns empty, WatchDirectory falls back to polling.
func TestWsPush_FallsBackToPollingWhenNoServerName(t *testing.T) {
	sess := new(MockWsSession)
	fs := filesystem.NewFileSystem(sess)

	sess.On("GetWsUrl").Return("wss://test.example.com/ws")
	sess.On("GetToken").Return("test-token")
	sess.On("GetMcpServerForTool", "get_file_change").Return("")

	pollCount := 0
	var mu sync.Mutex
	sess.On("CallMcpTool", "get_file_change", map[string]string{"path": "/tmp/poll"}).
		Return(&models.McpToolResult{Success: true, Data: "[]"}, nil).
		Run(func(args mock.Arguments) {
			mu.Lock()
			pollCount++
			mu.Unlock()
		})

	stopCh := make(chan struct{})
	wg, readyCh := fs.WatchDirectory("/tmp/poll", func([]*filesystem.FileChangeEvent) {}, 50*time.Millisecond, stopCh)
	<-readyCh

	time.Sleep(150 * time.Millisecond)
	close(stopCh)
	wg.Wait()

	mu.Lock()
	assert.GreaterOrEqual(t, pollCount, 2, "Should have polled at least twice")
	mu.Unlock()
}

// TestWsPush_FallsBackToPollingOnGetWsClientError verifies that when
// GetWsClient returns an error, WatchDirectory falls back to polling.
func TestWsPush_FallsBackToPollingOnGetWsClientError(t *testing.T) {
	sess := new(MockWsSession)
	fs := filesystem.NewFileSystem(sess)

	sess.On("GetWsUrl").Return("wss://test.example.com/ws")
	sess.On("GetToken").Return("test-token")
	sess.On("GetMcpServerForTool", "get_file_change").Return("wuying_filesystem")
	sess.On("GetWsClient").Return(nil, fmt.Errorf("connection refused"))

	pollCount := 0
	var mu sync.Mutex
	sess.On("CallMcpTool", "get_file_change", map[string]string{"path": "/tmp/poll"}).
		Return(&models.McpToolResult{Success: true, Data: "[]"}, nil).
		Run(func(args mock.Arguments) {
			mu.Lock()
			pollCount++
			mu.Unlock()
		})

	stopCh := make(chan struct{})
	wg, readyCh := fs.WatchDirectory("/tmp/poll", func([]*filesystem.FileChangeEvent) {}, 50*time.Millisecond, stopCh)
	<-readyCh

	time.Sleep(200 * time.Millisecond)
	close(stopCh)
	wg.Wait()

	mu.Lock()
	assert.GreaterOrEqual(t, pollCount, 2, "Should have polled in fallback mode")
	mu.Unlock()
}

// TestWsPush_ExistingPollingTestStillWorks ensures the old polling-only mock
// session (without wsProvider) continues to work.
func TestWsPush_ExistingPollingTestStillWorks(t *testing.T) {
	oldSess := new(MockWatchSession)
	fs := filesystem.NewFileSystem(oldSess)

	callbackCalled := false
	var mu sync.Mutex
	callback := func(events []*filesystem.FileChangeEvent) {
		mu.Lock()
		callbackCalled = true
		mu.Unlock()
	}

	oldSess.On("CallMcpTool", "get_file_change", map[string]string{"path": "/tmp/old"}).
		Return(&models.McpToolResult{
			Success: true,
			Data:    `[{"eventType":"create","path":"/tmp/old/f.txt","pathType":"file"}]`,
		}, nil)

	stopCh := make(chan struct{})
	wg, readyCh := fs.WatchDirectory("/tmp/old", callback, 50*time.Millisecond, stopCh)
	<-readyCh

	time.Sleep(120 * time.Millisecond)
	close(stopCh)
	wg.Wait()

	mu.Lock()
	assert.True(t, callbackCalled, "Callback should have been called in polling mode")
	mu.Unlock()
}

File: typescript/tests/unit/filesystem-watch-ws-push.test.ts
import { FileSystem, FileChangeEvent } from "../../src/filesystem/filesystem";
import * as sinon from "sinon";

function makeMockController() {
  const ctrl: any = {
    signal: {
      aborted: false,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
      onabort: null,
    } as any,
    abort: jest.fn(),
  };
  ctrl.abort.mockImplementation(() => {
    ctrl.signal.aborted = true;
    // Invoke all registered abort listeners
    for (const call of ctrl.signal.addEventListener.mock.calls) {
      if (call[0] === "abort" && typeof call[1] === "function") {
        try {
          call[1]();
        } catch (_e) {
          /* */
        }
      }
    }
  });
  return ctrl;
}

function makeMockSession(opts?: {
  wsUrl?: string;
  token?: string;
  serverName?: string | null;
}) {
  const sandbox = sinon.createSandbox();
  const session: any = {
    getAPIKey: sandbox.stub().returns("test-key"),
    getSessionId: sandbox.stub().returns("test-session"),
    callMcpTool: sandbox.stub().resolves({
      success: true,
      requestId: "r1",
      data: JSON.stringify([]),
      errorMessage: "",
    }),
    wsUrl: opts?.wsUrl ?? "wss://example.com/ws",
    token: opts?.token ?? "test-token",
    linkUrl: "",
    mcpTools: [],
    getMcpServerForTool: sandbox
      .stub()
      .returns(
        opts?.serverName === undefined ? "wuying_filesystem" : opts.serverName
      ),
    _isExpired: sandbox.stub().returns(false),
    getWsClient: sandbox.stub(),
  };
  return { session, sandbox };
}

function makeMockWsClient() {
  const registered: { target: string; callback: Function }[] = [];
  const stateListeners: Function[] = [];

  const ws: any = {
    registerCallback: jest.fn((target: string, cb: Function) => {
      registered.push({ target, callback: cb });
      return () => {
        const idx = registered.findIndex((r) => r.callback === cb);
        if (idx >= 0) registered.splice(idx, 1);
      };
    }),
    onConnectionStateChange: jest.fn((listener: Function) => {
      stateListeners.push(listener);
      return () => {
        const idx = stateListeners.indexOf(listener);
        if (idx >= 0) stateListeners.splice(idx, 1);
      };
    }),
    callStream: jest.fn().mockResolvedValue({
      invocationId: "inv-1",
      waitEnd: jest.fn().mockResolvedValue({ phase: "end" }),
      cancel: jest.fn(),
      write: jest.fn(),
    }),
    sendMessage: jest.fn().mockResolvedValue(undefined),
    connect: jest.fn().mockResolvedValue(undefined),
    _registered: registered,
    _stateListeners: stateListeners,
  };
  return ws;
}

describe("FileSystem watchDirectory WS Push Tests", () => {
  describe("Mode Selection", () => {
    it("should select WS push when all prerequisites met", () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      controller.abort();
      sandbox.restore();
    });

    it("should fall back to polling when no wsUrl", async () => {
      const { session, sandbox } = makeMockSession({ wsUrl: "" });
      const fs = new FileSystem(session);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        50,
        controller.signal
      );
      await ready;
      controller.abort();
      await monitoring;

      expect(session.getWsClient.called).toBe(false);
      sandbox.restore();
    });

    it("should fall back to polling when no server name", async () => {
      const { session, sandbox } = makeMockSession({ serverName: null });
      const fs = new FileSystem(session);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        50,
        controller.signal
      );
      await ready;
      controller.abort();
      await monitoring;

      expect(session.getWsClient.called).toBe(false);
      sandbox.restore();
    });
  });

  describe("Subscription", () => {
    it("should call subscribe_file_change via callStream", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/home/user/project",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));
      controller.abort();
      await monitoring;

      expect(mockWs.callStream).toHaveBeenCalled();
      const callArgs = mockWs.callStream.mock.calls[0][0];
      expect(callArgs.target).toBe("wuying_filesystem");
      expect(callArgs.data.method).toBe("subscribe_file_change");
      expect(callArgs.data.params.path).toBe("/home/user/project");
      sandbox.restore();
    });

    it("should register callback with server name", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));
      controller.abort();
      await monitoring;

      expect(mockWs.registerCallback).toHaveBeenCalled();
      const regArgs = mockWs.registerCallback.mock.calls[0];
      expect(regArgs[0]).toBe("wuying_filesystem");
      expect(typeof regArgs[1]).toBe("function");
      sandbox.restore();
    });

    it("should register connection state change listener", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));
      controller.abort();
      await monitoring;

      expect(mockWs.onConnectionStateChange).toHaveBeenCalled();
      const stateArgs = mockWs.onConnectionStateChange.mock.calls[0];
      expect(typeof stateArgs[0]).toBe("function");
      sandbox.restore();
    });
  });

  describe("Push Event Dispatch", () => {
    it("should dispatch push events to user callback", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const received: FileChangeEvent[] = [];
      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        (events) => received.push(...events),
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));

      const pushHandler = mockWs._registered[0]?.callback;
      expect(pushHandler).toBeDefined();

      pushHandler({
        data: {
          eventType: "file_change",
          path: "/watch",
          events: [
            { eventType: "create", path: "/watch/new.txt", pathType: "file" },
            { eventType: "modify", path: "/watch/old.py", pathType: "file" },
          ],
        },
      });

      expect(received).toHaveLength(2);
      expect(received[0].eventType).toBe("create");
      expect(received[0].path).toBe("/watch/new.txt");
      expect(received[1].eventType).toBe("modify");

      controller.abort();
      await monitoring;
      sandbox.restore();
    });

    it("should ignore push events for different path", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const received: FileChangeEvent[] = [];
      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        (events) => received.push(...events),
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));

      const pushHandler = mockWs._registered[0]?.callback;
      pushHandler({
        data: {
          eventType: "file_change",
          path: "/other",
          events: [
            { eventType: "create", path: "/other/x.txt", pathType: "file" },
          ],
        },
      });

      expect(received).toHaveLength(0);

      controller.abort();
      await monitoring;
      sandbox.restore();
    });

    it("should ignore push events with wrong eventType", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const received: FileChangeEvent[] = [];
      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        (events) => received.push(...events),
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));

      const pushHandler = mockWs._registered[0]?.callback;
      pushHandler({
        data: {
          eventType: "other_type",
          path: "/watch",
          events: [
            { eventType: "create", path: "/watch/x.txt", pathType: "file" },
          ],
        },
      });

      expect(received).toHaveLength(0);

      controller.abort();
      await monitoring;
      sandbox.restore();
    });
  });

  describe("Unsubscribe on Stop", () => {
    it("should send unsubscribe_file_change on stop", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));
      controller.abort();
      await monitoring;

      expect(mockWs.sendMessage).toHaveBeenCalledWith("wuying_filesystem", {
        method: "unsubscribe_file_change",
        params: { path: "/watch" },
      });
      sandbox.restore();
    });

    it("should unregister push callback on stop", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));

      expect(mockWs._registered).toHaveLength(1);

      controller.abort();
      await monitoring;

      expect(mockWs._registered).toHaveLength(0);
      sandbox.restore();
    });

    it("should unregister state listener on stop", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      session.getWsClient.resolves(mockWs);

      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        () => {},
        500,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));

      expect(mockWs._stateListeners).toHaveLength(1);

      controller.abort();
      await monitoring;

      expect(mockWs._stateListeners).toHaveLength(0);
      sandbox.restore();
    });
  });

  describe("Fallback to Polling", () => {
    it("should fall back to polling when subscribe fails", async () => {
      const { session, sandbox } = makeMockSession();
      const fs = new FileSystem(session);
      const mockWs = makeMockWsClient();
      mockWs.callStream.mockRejectedValue(new Error("subscribe failed"));
      session.getWsClient.resolves(mockWs);

      let pollCount = 0;
      session.callMcpTool.callsFake(() => {
        pollCount++;
        if (pollCount > 3) {
          return Promise.resolve({
            success: true,
            requestId: "r",
            data: JSON.stringify([
              { eventType: "create", path: "/watch/new.txt", pathType: "file" },
            ]),
            errorMessage: "",
          });
        }
        return Promise.resolve({
          success: true,
          requestId: "r",
          data: JSON.stringify([]),
          errorMessage: "",
        });
      });

      const received: FileChangeEvent[] = [];
      const controller = makeMockController();
      const { monitoring, ready } = fs.watchDirectory(
        "/watch",
        (events) => received.push(...events),
        50,
        controller.signal
      );
      await ready;
      await new Promise((r) => setTimeout(r, 500));
      controller.abort();
      await monitoring;

      expect(pollCount).toBeGreaterThan(3);
      expect(received.length).toBeGreaterThan(0);
      sandbox.restore();
    });
  });
});
